gensim/similarities/nmslib.py

```diff
@@ -58,28 +95,16 @@ def __init__(self, model=None, index_params=None, query_time_params=None):
         ----------
         model : :class:`~gensim.models.base_any2vec.BaseWordEmbeddingsModel`, optional
             Model, that will be used as source for index.
+            If the model is None, index and labels are not initialized.
+            In that case please load or init the index and labels by yourself.
```

Comment:
This is better than before, but still not very helpful. The main questions: How is the user supposed to load/init the index? Which functions to call?

Next, does it make sense to pass model=None from an API perspective? Does it allow the user to do something new? If it just means the user has to do more work using other functions, and does not have any advantages, then I think we should force them to pass a non-None model, because this aligns with the Zen of Python: there should only be one obvious way to do things.

Reply:
In the load function, we call this init function with model=None.
```
    @classmethod
    def load(cls, fname):
        """Load a NmslibIndexer instance from a file

        Parameters
        ----------
        fname : str
            Path to dump with NmslibIndexer.

        """
        fname_dict = fname + '.d'
        with open(fname_dict, 'rb') as f:
            d = _pickle.load(f)
        index_params = d['index_params']
        query_time_params = d['query_time_params']
        nmslib_instance = cls(index_params=index_params, query_time_params=query_time_params)
        index = nmslib.init()
        index.loadIndex(fname)
        nmslib_instance.index = index
        nmslib_instance.labels = d['labels']
        return nmslib_instance
```
Do you think this init function should not allow model=None? In that case, this load function should have a model argument? (i.e, load(cls, fname, model))

Reply:
OK, so we have two separate use cases for the initializer.

1. Construct a new class, passing a model (called by user)
2. Construct a new class, don't pass a model (load function, internal use case)

So it sounds like for the user, they will **always** have to pass a model. So the model should not be a keyword parameter. Is that correct? If so, let's update the initializer parameters and docstring to reflect that. Next, to support the second use case, we should make the load function pass None as the model explicitly.

Next, I think the second use case above isn't clearly documented. Look at the save/load example:

```
    >>> indexer = NmslibIndexer(model)
    >>> _, temp_fn = mkstemp()
    >>> indexer.save(temp_fn)
    >>>
    >>> new_indexer = NmslibIndexer.load(temp_fn)
    >>> model.most_similar("cat", topn=2, indexer=new_indexer)
    [('cat', 1.0), ('meow', 0.5595494508743286)]
```

When `new_indexer` is created, it knows nothing about the model. The docstring for the initializer says this:

```
            If the model is None, index and labels are not initialized.
            In that case please load or init the index and labels by yourself.
```

Why is the code not loading/initializing index and labels?

Reply:
I see. I will change the initializer's parameters and docstring so that the model should not be a keyword parameter. And I will make the load function pass None as the model explicitly.

In the load function, we create an indexer with model=None. After that, we assign index and labels.
```
        nmslib_instance = cls(index_params=index_params, query_time_params=query_time_params)
        index = nmslib.init()
        index.loadIndex(fname)
        nmslib_instance.index = index
        nmslib_instance.labels = d['labels']
        return nmslib_instance
```
So when we use load function to create a indexer, the indexer has index and labels.

Reply:
Yes. We should also move these lines:

```
            If the model is None, index and labels are not initialized.
            In that case please load or init the index and labels by yourself.
```

from the docstring to somehere around [here](https://github.com/RaRe-Technologies/gensim/blob/7e2d07fcf13f1162b5157a331ee5d385419f7c13/gensim/similarities/nmslib.py#L117). This sort of information is purely an implementation detail, so should not be visible to users.

Reply:
OK, thanks. I will move it.

Reply:
Fixed those.